### PR TITLE
[DefaultType] fix rigidcoord compilation

### PR DESCRIPTION
**Pull request by bakpaul**
______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Review by alxbilger — APPROVED** [review]
Thanks for your contribution. I wonder how it was not detected before

**Comment by fredroy** [comment]
I dont think the fix is good ; the "toMatrix()" is available for Mat3x3 but not for Mat4x4 (there was a ` toMatrix`  for Mat4x4 but it was deleted and replace with `toHomogenousMatrix()` ).
This RigidCoord::toMatrix() is templated on the Mat type actually, and is really ill-formed (no assert on the size of the given matrix, etc).
I think the solution is to either:
 - test if the matrix is 3x3 or 4x4 (and failed to compile with a message with a static assert) and calling the respective ` toMatrix()`  or ` toHomogenousMatrix ()` 
 - or remove the template, force it to Mat3x3 and create an other `toHomogenousMatrix` in RigidCoord. Consisntent with Quat but duplicating code and breaking.

I am more for the first solution.

> Thanks for your contribution. I wonder how it was not detected before

I guess this was not detected before because nobody was calling RigidCoord::toMatrix with a Mat4x4 with the code compiled on the CI. Actually, nobody calls RigidCoord::toMatrix, either with mat3 or mat4  apparently.

File: Sofa/framework/DefaultType/src/sofa/defaulttype/RigidCoord.h [hunk 1]
```diff
@@ -269,7 +269,7 @@ class RigidCoord<3,real>
     void toMatrix( Mat& m) const
     {
         m.identity();
-        orientation.toMatrix(m);
+        orientation.toHomogeneousMatrix(m);
```
Comment (fredroy) [inline Sofa/framework/DefaultType/src/sofa/defaulttype/RigidCoord.h]:
```suggestion
       if constexpr (std::is_same_v<Mat, sofa::type::Mat<3,3,real>)
       {
            orientation.toMatrix(m);
       }
       else if constexpr (std::is_same_v<Mat, sofa::type::Mat<4,4,real>)
       {
           orientation.toHomogeneousMatrix(m);
       }
       else
       {
           static_assert(false, "toMatrix() only supports sofa::type::Mat<3,3,Real> or sofa::type::Mat<4,4,Real>");
       }
```

YOLO, I did not try to compile it so typos may exist 🤫

**Review by fredroy — CHANGES_REQUESTED** [review]

File: Sofa/framework/DefaultType/src/sofa/defaulttype/RigidCoord.h (hunk 1 above)
Comment (bakpaul) [reply Sofa/framework/DefaultType/src/sofa/defaulttype/RigidCoord.h]: I did the modifications with some fix.

**Review by alxbilger — CHANGES_REQUESTED** [review]
If only 2 Mat types are allowed, I suggest to define 2 different overloads:

```cpp
void toMatrix( sofa::type::Mat<3,3,real>& m) const
{
...
}

void toMatrix( sofa::type::Mat<4,4,real>& m) const
{
...
}
```

Also, check if the 2d rigid needs to be changed as well.

A unit test would be welcomed as well. I think this template function is not used at all in SOFA. That's why this problem was not detected.

**Comment by alxbilger** [comment]
The changes I suggested are here: https://github.com/mimesis-inria/sofa/pull/18

**Pull request closed without merging**